src/ralph/virtual/management/commands/openstack_sync.py

```diff
@@ -361,6 +368,75 @@ def _get_hypervisor(host_name, server_id):
             ))
             return None
 
+    def _match_physical_and_cloud_hosts(self):
+        """Connect CloudHosts and DC assets according to data from Ironic."""
+
+        provider = getattr(
+            settings, 'OPENSTACK_IRONIC_PROVIDER', 'openstack-ironic'
+        )
+        os_conf = next((conf for conf in settings.OPENSTACK_INSTANCES
+                        if conf['provider'] == provider), None)
+
+        if os_conf is None:
+            logger.error('Ironic is not configured.')
+            return
+
+        ironic_client = get_ironic_client(
+            api_version=os_conf['version'],
+            os_username=os_conf['username'],
+            os_password=os_conf['password'],
+            os_tenant_name=os_conf['tenant_name'],
+            os_auth_url=os_conf['auth_url']
+        )
+
+        nodes = ironic_client.node.list(
+            associated=True,
+            fields=['extra', 'instance_uuid']
+        )
+
+        not_found_message_tpl = (
+            '{} with the host id or serial number {} was not found. Check if '
+            'Ralph is synchronized with OpenStack or add it manually.'
+        )
+
+        for node in nodes:
+            try:
+                host = CloudHost.objects.get(host_id=node.instance_uuid)
+                asset = DataCenterAsset.objects.get(
+                    sn=node.extra['serial_number']
+                )
+            except DataCenterAsset.DoesNotExist:
+                logger.warning(
+                    not_found_message_tpl.format(
+                        'DC asset',
+                        node.extra['serial_number']
+                    )
+                )
+            except CloudHost.DoesNotExist:
+                logger.warning(
+                    not_found_message_tpl.format(
+                        'Cloud host',
+                        node.instance_uuid
+                    )
+                )
+            except DataCenterAsset.MultipleObjectsReturned:
+                logger.error(
```

Comment:
fyi: https://github.com/allegro/ralph/blob/ng/src/ralph/assets/models/assets.py#L349 - unless `node.extra['serial_number']` is None, this case will never occur (sn is unique)